### PR TITLE
27 binary to decimal converter

### PR DESCRIPTION
**Pull request by Khusro-S**
---
name: solution review 
about: A template PR for code review with a checklist
---

### Files

- [x] The file name describes the function's behavior
- [x] There is a module docstring in the function file
- [x] The test file's name matches the function file name -
  `/tests/test_file_name.py`
- [x] There is a module docstring in the tests file

### Unit Tests

- [x] The test class has a helpful name in PascalCase
- [x] The test class has a docstring
- Every unit test has
  - [x] A helpful name
  - [x] A clear docstring
  - [x] Only one assertion
  - [x] There is no logic in the unit test
- [x] All tests pass
- [x] There are tests for defensive assertions
- [x] There are tests for boundary cases

### Function Docstring

- [x] The function's behavior is described
- The function's arguments are described:
  - [x] Type
  - [x] Purpose
  - [x] Other assumptions (eg. if it's a number, what's the expected range?)
- The return value is described
  - [x] Type
  - [x] Other assumptions are documented
- The defensive assertions are documented using `Raises:`
  - [x] Each assumption about an argument is checked with an assertion
  - [x] Each assertion checks for _only one_ assumption about the argument
- [x] Include 3 or more (passing!) doctests

### The Function

- [x] The function's name describes it's behavior
- [x] The function's name matches the file name
- [x] The function has correct type annotations

### Do's

- [x] Variable names help to understand the strategy
- [x] Any comments are clear and describe the strategy
- [x] Lines of code are spaced to help show different stages of the strategy

### Don'ts

- [x] The function's strategy _is not_ described in the documentation
- [x] Comments explain the _strategy_, **not** the _implementation_
- [x] The function _does not_ have more comments than code
  - If it does, consider finding a new strategy or a simpler implementation

## Implementation

- [x] The code passes the formatting checks
- [x] The code passes all Ruff linting checks
- [x] The code has no (reasonable) Pylint errors
  - In code review, you can decide when fixing a Pylint error is helpful and
    when it's too restricting.
- [x] Variables are named with snake_case
- [x] Variable names are clear and helpful
- [x] The code follows the strategy as simply as possible
- [x] The implementation is as simple as possible given the strategy
- [x] The code includes defensive assertions
- [x] Defensive assertions include as little logic as possible

### CODE REVIEW

File: solutions/binary_to_decimal.py [hunk 1]
```diff
@@ -0,0 +1,53 @@
+#!/usr/bin/env python3
+# -*- coding: utf-8 -*-
+"""
+A module for converting binary numbers to decimal.
+
+Module contents:
+    - binary_to_decimal: converts binary numbers to decimal.
+
+Created on XX XX XX
```
Comment (Mohamed-Elnageeb) [inline solutions/binary_to_decimal.py:9]: add date

File: solutions/binary_to_decimal.py [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+@author: Khusro Sakhi
+"""
+
+
+def binary_to_decimal(binary_str) -> int:
```
Comment (Mohamed-Elnageeb) [inline solutions/binary_to_decimal.py:14]: input type missing

File: solutions/tests/test_binary_to_decimal.py [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+
+Test categories:
+  - Standard cases: Typical binary numbers with various lengths and values.
+  - Edge cases: Binary strings with leading zeros, empty strings, or single digits.
+  - Defensive tests: Tests for invalid input types (e.g., non-strings, floats, `None`)
+    and assertions (e.g., strings containing non-binary characters).
+
+
+
+Created on XX XX XX
```
Comment (Mohamed-Elnageeb) [inline solutions/tests/test_binary_to_decimal.py:15]: add date

File: solutions/tests/test_binary_to_decimal.py [hunk 4]
```diff
@@ -0,0 +1,90 @@
+#!/usr/bin/env python3
+# -*- coding: utf-8 -*-
+"""
+Test module for binary_to_decimal function.
+Contains intentionally buggy tests for debugging practice.
```
Comment (Mohamed-Elnageeb) [inline solutions/tests/test_binary_to_decimal.py:5]: couldn,t find the buggy test

**Review by Mohamed-Elnageeb — COMMENTED** [review]
nice work Khusro, code implementation is clear , well formated and its strategy is self-explanatory. 👍🤩
A few tiny changes suggested as comments

File: solutions/binary_to_decimal.py (hunk 1 above)
Comment (Khusro-S) [reply solutions/binary_to_decimal.py:9]: I'm on it 🫡

File: solutions/binary_to_decimal.py (hunk 2 above)
Comment (Khusro-S) [reply solutions/binary_to_decimal.py:14]: I guess I missed it everywhere 😅

File: solutions/tests/test_binary_to_decimal.py (hunk 3 above)
Comment (Khusro-S) [reply solutions/tests/test_binary_to_decimal.py:15]: Got it 👍

File: solutions/tests/test_binary_to_decimal.py (hunk 4 above)
Comment (Khusro-S) [reply solutions/tests/test_binary_to_decimal.py:5]: I guess I should just remove this comment 🤔

**Comment by Khusro-S** [comment]
Thank you, will make the suggested changes 🙌

**Comment by Zeinab15** [comment]
Hello @Khusro-S ,
 I reviewed it, everything is good  just update it as Mohamed mentioned.
 All the best for you.

**Comment by Khusro-S** [comment]
Thank you both @Mohamed-Elnageeb and @Zeinab15. I  made all the suggested changes and corrections 🙌

**Review by Zeinab15 — APPROVED** [review]
Well done  @Khusro-S, I approve

**Review by Mohamed-Elnageeb — APPROVED** [review]
perfectly done!

**Review by Zeinab15 — APPROVED** [review]

**Review by Mohamed-Elnageeb — APPROVED** [review]
Well done
